witchcraft/witchcraft.go

```diff
@@ -601,17 +603,42 @@ func (s *Server) Start() (rErr error) {
 	ctx = metrics.WithRegistry(ctx, metricsRegistry)
 
 	// initialize loggers
-	s.initLoggers(baseInstallCfg.UseConsoleLog, wlog.InfoLevel, metricsRegistry)
-
-	// add loggers to context
-	ctx = s.withLoggers(ctx)
+	s.initLoggers(baseInstallCfg.UseConsoleLog, wlog.InfoLevel, metricsRegistry, nil)
 
 	// load runtime configuration
 	baseRefreshableRuntimeCfg, refreshableRuntimeCfg, configReloadHealthCheckSource, err := s.initRuntimeConfig(ctx)
 	if err != nil {
 		return err
 	}
 
+	// extract the TCP JSON receiver from the runtime config
+	servicesCfg := baseRefreshableRuntimeCfg.CurrentBaseRuntimeConfig().ServiceDiscovery
+	receiverCfg, err := servicesCfg.MustClientConfig("sls-log-tcp-json-receiver")
+	if err == nil && len(receiverCfg.URIs) > 0 {
+		tlsConfig, err := tlsconfig.NewClientConfig(
+			tlsconfig.ClientRootCAFiles(receiverCfg.Security.CAFiles...),
+			tlsconfig.ClientKeyPairFiles(receiverCfg.Security.CertFile, receiverCfg.Security.KeyFile),
+		)
+		if err != nil {
+			return err
+		}
+		connProvider, err := tcpjson.NewTCPConnProvider(receiverCfg.URIs, tlsConfig)
+		if err != nil {
+			return err
+		}
+		envelopeMetadata := tcpjson.GetEnvelopeMetadata()
+		tcpWriter := tcpjson.NewTCPWriter(envelopeMetadata, connProvider)
+		defer func() {
+			_ = tcpWriter.Close()
+		}()
+		// re-initialize the loggers with the TCP writer
+		s.initLoggers(baseInstallCfg.UseConsoleLog, wlog.InfoLevel, metricsRegistry, tcpWriter)
```

Comment:
Do we know that running this again is safe? Would it make more sense to do the TCP writer logic first and then pass `tcpWriter` to the place where `initLoggers` is currently called?

Reply:
Ah OK I see, this requires the runtime config to be loaded first. Then would just want to double-check the contract for the init loggers call.

Reply:
Yep that is the correct observation. The `initLoggers` call only operates over the logger objects, so it's safe to re-run this again given we shouldn't have to worry about concurrent access or anything getting clobbered. However, you did point out a bug that I expanded on below.